File: stack/C/stack.c
#include <stdio.h>
#include <errno.h>
#include <stdlib.h>

typedef enum{false, true} bool;

typedef struct
{
	int size, top;
	int *STACK;
}stack;

void init(stack* s, int SIZE)
{
	s->size = SIZE;
	s->STACK = (int*)malloc(SIZE * sizeof(int));
	if(s->STACK == NULL){
		perror("Memory allocation failed.");
	}
	s->top = -1;
}

bool isEmpty(stack* s)
{
	if(s->top == -1)
		return true;
	else
		return false;
}

bool isFull(stack* s)
{
	if(s->top == s->size - 1)
		return true;
	else
		return false;
}

void push(stack* s, int item)
{
	if(isFull(s)){
		perror("Stack overflow\n");
		exit(1);
		}
	else{
		s->top++;
		s->STACK[s->top] = item;
	}
}

int pop(stack* s)
{
	if(isEmpty(s)){
		perror("Stack underflow\n");
		exit(1);
	}
	else{
		int item = s->STACK[s->top];
		s->top--;
		return item;
	}
}

void display(stack* s)
{
	printf("Items in the stack: \n");
	int i = s->top;
	while(i!=-1){
		printf("%d\n",s->STACK[i]);
		i--;
	}
}

void destroy(stack* s)
{
	free(s->STACK);
}

int main()
{
	stack s;
	init(&s, 5);
	push(&s, 5);
	push(&s, 2);
	push(&s, 1);
	display(&s);
	destroy(&s);
}

File: stack/zig/stack.zig
const std = @import("std");

const Stack = struct {
   size: usize,
   top: isize,
   STACK: []i32,
   
   pub fn init(SIZE: usize) !Stack {
       const allocator = std.heap.page_allocator;
       const stack_memory = try allocator.alloc(i32, SIZE);
       
       return Stack{
           .size = SIZE,
           .top = -1,
           .STACK = stack_memory,
       };
   }
   
   pub fn push(self: *Stack, item: i32) void {
       if (self.top >= @intCast(isize, self.size) - 1) {
           std.debug.print("Stack overflow\n", .{});
           return;
       }
       self.top += 1;
       self.STACK[@intCast(usize, self.top)] = item;
   }
   
   pub fn pop(self: *Stack) ?i32 {
       if (self.top == -1) {
           std.debug.print("Stack underflow\n", .{});
           return null;
       }
       const item = self.STACK[@intCast(usize, self.top)];
       self.top -= 1;
       return item;
   }
   
   pub fn display(self: *Stack) void {
       std.debug.print("Items in the stack: ", .{});
       var i: usize = 0;
       while (i <= @intCast(usize, self.top)) : (i += 1) {
           std.debug.print("{d} ", .{self.STACK[i]});
       }
       std.debug.print("\n", .{});
   }
};

pub fn main() !void {
   var s = try Stack.init(5);
   std.debug.print("The size of stack is {d}\n", .{s.size});
   s.push(5);
   s.push(2);
   s.push(1);
   s.display();
}
